apps/passport-server/src/database/queries/rateLimit.ts

```diff
@@ -0,0 +1,20 @@
+import { Pool } from "postgres-pool";
+import { RateLimit } from "../models";
+import { sqlQuery } from "../sqlQuery";
+
+export async function checkRateLimit(
+  client: Pool,
+  eventType: RateLimit["type"],
+  eventId: string
+): Promise<boolean> {
+  return (
+    await sqlQuery(client, `SELECT take_token($1, $2)`, [eventType, eventId])
```

Comment:
What happens if the database query fails here?  Exception because rows[0] doesn't exist?  Should there be a retry?  Is there a risk that it uses up your rate limit?
It might be nice if what the user sees in case of such a failure is different from being over their limit.

Reply:
If this fails then it throws an exception. What the user sees depends on what the calling code does with the exception.

`sqlQuery` does some automatic retries, to cover the case where we have stale connections in the pool. Beyond that, if the database server is not reachable then there is a limit to what we can do at this layer. At the higher level we might want to do in-memory rate limiting to prevent denial of service attacks.

Reply:
suggestion: It may be nice to parse out and throw a new type of exception - `RateLimitException`, so that we get more visibility into rate limit errors in our various logging and error reporting tools.

Reply:
I've added extra logging in the `RateLimitService`, which I think covers the logging need. The question of whether to add an exception comes down to whether we want usage to be like this:
```typescript
try {
  rateLimitService.requestRateLimitedAction("SOME_ACTION_TYPE", "someId");
  // perform the action
}
catch (e) {
  // handle the exception
  // return 401 or 429 error
}
```
or:
```typescript
if (!rateLimitService.requestRateLimitedAction("SOME_ACTION_TYPE", "someId")) {
  // return a 401 or 429 error
} else {
  // perform the action
}
```

I don't really mind which, so I'm happy to go with either if this maintains consistency with a design pattern.